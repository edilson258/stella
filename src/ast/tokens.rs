#![allow(dead_code)]

use serde::{Deserialize, Serialize};

use crate::utils::location::Location;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Token {
  pub kind: TokenKind,
  pub location: Location,
}

impl Token {
  pub fn new(kind: TokenKind, location: Location) -> Token {
    Token { kind, location }
  }

  pub fn new_end_of_file(location: Location) -> Token {
    Token::new(TokenKind::EOF, location)
  }

  pub fn new_identifier(location: Location, identifier: String) -> Token {
    Token::new(TokenKind::Identifier(identifier), location)
  }

  pub fn new_number(location: Location, number: f64) -> Token {
    Token::new(TokenKind::Number(number), location)
  }

  pub fn new_string(location: Location, string: String) -> Token {
    Token::new(TokenKind::String(string), location)
  }

  pub fn new_keyword_or_identifier(location: Location, keyword: String) -> Token {
    match keyword.as_str() {
      "function" => Token::new(TokenKind::Function, location),
      "local" => Token::new(TokenKind::Local, location),
      "if" => Token::new(TokenKind::If, location),
      "then" => Token::new(TokenKind::Then, location),
      "else" => Token::new(TokenKind::Else, location),
      "elseif" => Token::new(TokenKind::ElseIf, location),
      "end" => Token::new(TokenKind::End, location),
      "while" => Token::new(TokenKind::While, location),
      "do" => Token::new(TokenKind::Do, location),
      "for" => Token::new(TokenKind::For, location),
      "in" => Token::new(TokenKind::In, location),
      "repeat" => Token::new(TokenKind::Repeat, location),
      "until" => Token::new(TokenKind::Until, location),
      "return" => Token::new(TokenKind::Return, location),
      "break" => Token::new(TokenKind::Break, location),
      "true" => Token::new(TokenKind::True, location),
      "false" => Token::new(TokenKind::False, location),
      "nil" => Token::new(TokenKind::Nil, location),
      "type" => Token::new(TokenKind::Type, location),
      "enum" => Token::new(TokenKind::Enum, location),
      _ => Token::new(TokenKind::Identifier(keyword), location),
    }
  }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum TokenKind {
  // keywords
  Function,
  Local,
  If,
  Then,
  Else,
  ElseIf,
  End,
  While,
  Do,
  For,
  In,
  Repeat,
  Until,
  Return,
  Break,
  True,
  False,
  Nil,
  Type,
  Enum,

  // Operators and Delimiters
  Tilde,        // ~
  Plus,         // +
  Minus,        // -
  Star,         // *
  StarAssign,   // *=
  Slash,        // /
  SlashAssign,  // /=
  Percent,      // %
  Equal,        // ==
  NotEqual,     // ~=
  Less,         // <
  Greater,      // >
  LessEqual,    // <=
  GreaterEqual, // >=
  Assign,       // =
  PlusAssign,   // +=
  MinusAssign,  // -=
  And,          // and
  Or,           // or
  Not,          // not
  Hash,         // #
  Comma,        // ,
  Semicolon,    // ;
  Colon,        // :
  DoubleColon,  // ::
  Dot,          // .
  DoubleDot,    // ..
  TripleDot,    // ...
  LeftParen,    // (
  RightParen,   // )
  LeftBrace,    // {
  RightBrace,   // }
  LeftBracket,  // [
  RightBracket, // ]

  // Literals
  Identifier(String),
  Number(f64),
  String(String),

  // Outros
  EOF, // End of file
}
