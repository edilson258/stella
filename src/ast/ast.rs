#![allow(dead_code)]
use super::tokens::Token;
use crate::{
  types::Type,
  utils::range::{create_middle_range, Range},
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Program {
  pub statements: Vec<Statement>,
}

impl Program {
  pub fn new() -> Program {
    Program { statements: Vec::new() }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Statement {
  Assign(AssignStatement),
  Function(FunctionStatement),
  Return(ReturnStatement),
  If(IfStatement),
  While(WhileStatement),
  Repeat(RepeatStatement),
  For(ForStatement),
  Break(BreakStatement),
  Goto(GotoStatement),
  Block(BlockStatement),
  Empty(EmptyStatement),
  VariableDeclaration(VariableDeclaration),
  TypeDeclaration(TypeDeclaration),
  Expression(Expression),
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AssignStatement {
  pub values: Vec<(Token, Option<Type>)>,
  pub value: Expression,
}

impl AssignStatement {
  pub fn new(values: Vec<(Token, Option<Type>)>, value: Expression) -> Self {
    AssignStatement { values, value }
  }

  pub fn get_range(&self) -> Range {
    let left_range = self.values.first().unwrap().0.range.clone();
    let right_range = self.value.get_range();
    create_middle_range(&left_range, &right_range)
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct FunctionStatement {
  pub name: Token,
  pub local: bool,
  pub arguments: Vec<(Token, Option<Type>)>,
  pub return_type: Option<Type>,
  pub generics: Vec<Type>,
  pub body: Box<Statement>,
  pub range: Range,
}

impl FunctionStatement {
  pub fn new(
    name: Token,
    local: bool,
    generics: Vec<Type>,
    arguments: Vec<(Token, Option<Type>)>,
    return_type: Option<Type>,
    body: Statement,
    range: Range,
  ) -> Self {
    FunctionStatement { name, local, generics, arguments, return_type, body: Box::new(body), range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ReturnStatement {
  pub values: Vec<Expression>,
  pub range: Range,
}

impl ReturnStatement {
  pub fn new(values: Vec<Expression>, range: Range) -> Self {
    ReturnStatement { values, range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct IfStatement {
  pub condition: Expression,
  pub then_body: Box<Statement>,
  pub else_body: Option<Box<Statement>>,
  pub range: Range,
}

impl IfStatement {
  pub fn new(condition: Expression, then_body: Statement, else_body: Option<Statement>, range: Range) -> Self {
    IfStatement { condition, then_body: Box::new(then_body), else_body: else_body.map(Box::new), range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WhileStatement {
  pub condition: Expression,
  pub body: Box<Statement>,
  pub range: Range,
}

impl WhileStatement {
  pub fn new(condition: Expression, body: Statement, range: Range) -> Self {
    WhileStatement { condition, body: Box::new(body), range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RepeatStatement {
  pub body: Box<Statement>,
  pub condition: Expression,
  pub range: Range,
}

impl RepeatStatement {
  pub fn new(body: Statement, condition: Expression, range: Range) -> Self {
    RepeatStatement { body: Box::new(body), condition, range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ForStatement {
  pub variable: Expression,
  pub init: Expression,
  pub limit: Expression,
  pub step: Option<Expression>,
  pub body: Box<Statement>,
  pub range: Range,
}

impl ForStatement {
  pub fn new(
    variable: Expression,
    init: Expression,
    limit: Expression,
    step: Option<Expression>,
    body: Statement,
    range: Range,
  ) -> Self {
    ForStatement { variable, init, limit, step, body: Box::new(body), range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BreakStatement {
  pub range: Range,
}

impl BreakStatement {
  pub fn new(range: Range) -> Self {
    BreakStatement { range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GotoStatement {
  pub label: Option<String>,
  pub range: Range,
}

impl GotoStatement {
  pub fn new(label: Option<String>, range: Range) -> Self {
    GotoStatement { label, range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BlockStatement {
  pub statements: Vec<Statement>,
  pub range: Range,
}

impl BlockStatement {
  pub fn new(statements: Vec<Statement>, range: Range) -> Self {
    BlockStatement { statements, range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct EmptyStatement {}

impl EmptyStatement {}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub enum TypeInitializer {
  Type(Type),
  Function(TypeFunction),
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub struct TypeDeclaration {
  pub name: Token,
  pub initiizer: Type,
  pub range: Range,
  pub generis: Vec<String>,
}

impl TypeDeclaration {
  pub fn new(name: Token, generis: Vec<String>, initiizer: Type, range: Range) -> Self {
    TypeDeclaration { name, generis, initiizer, range }
  }
}

#[derive(Debug, Serialize, Deserialize, PartialEq)]
pub struct TypeFunction {
  pub params: Vec<Type>,
  pub return_type: Type,
  pub range: Range,
}

impl TypeFunction {
  pub fn new(params: Vec<Type>, return_type: Type, range: Range) -> Self {
    TypeFunction { params, return_type, range }
  }

  pub fn get_range(&self) -> Range {
    self.range.clone()
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct VariableDeclaration {
  pub values: Vec<(Token, Option<Type>)>,
  pub local: bool,
  pub initializer: Option<Expression>,
}

impl VariableDeclaration {
  pub fn new(values: Vec<(Token, Option<Type>)>, local: bool, initializer: Option<Expression>) -> Self {
    VariableDeclaration { values, local, initializer }
  }

  pub fn get_range(&self) -> Range {
    let left_range = self.values.first().unwrap().0.range.clone();
    let right_range = self.initializer.as_ref().unwrap().get_range();
    create_middle_range(&left_range, &right_range)
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Expression {
  Literal(LiteralExpression),
  Identifier(Identifier),
  Call(CallExpression),
  Unary(UnaryExpression),
  Grouped(GroupedExpression),
  Binary(BinaryExpression),
  Require(RequireExpression),
  Function(FunctionExpression),
}

impl Expression {
  pub fn new_literal(value: LiteralExpression) -> Self {
    Expression::Literal(value)
  }

  pub fn new_identifier(name: String, range: Range) -> Self {
    Expression::Identifier(Identifier::new(name, range))
  }

  pub fn new_call(name: Token, args: Expression, range: Range) -> Self {
    Expression::Call(CallExpression::new(name, Box::new(args), range))
  }

  pub fn new_require(module_name: Token, range: Range) -> Self {
    Expression::Require(RequireExpression::new(module_name, range))
  }

  pub fn new_grouped(expressions: Vec<Expression>, range: Range) -> Self {
    Expression::Grouped(GroupedExpression::new(expressions, range))
  }

  pub fn new_function(
    arguments: Vec<(Token, Option<Type>)>,
    return_type: Option<Type>,
    body: Statement,
    loc: Range,
  ) -> Self {
    Expression::Function(FunctionExpression::new(arguments, return_type, body, loc))
  }

  pub fn get_range(&self) -> Range {
    match self {
      Expression::Literal(literal) => literal.get_range(),
      Expression::Identifier(identifier) => identifier.range.clone(),
      Expression::Call(call) => call.get_range(),
      Expression::Binary(binary) => binary.get_range(),
      Expression::Require(require) => require.get_range(),
      Expression::Grouped(grouped) => grouped.get_range(),
      Expression::Unary(unary) => unary.get_range(),
      Expression::Function(function) => function.get_range(),
    }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RequireExpression {
  pub module_name: Token,
  pub range: Range,
}

impl RequireExpression {
  pub fn new(module_name: Token, range: Range) -> Self {
    RequireExpression { module_name, range }
  }

  pub fn get_range(&self) -> Range {
    let right_range = self.module_name.range.clone();
    let left_range = self.range.clone();
    create_middle_range(&left_range, &right_range)
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GroupedExpression {
  pub expressions: Vec<Expression>,
  pub range: Range,
}

impl GroupedExpression {
  pub fn new(expressions: Vec<Expression>, range: Range) -> Self {
    GroupedExpression { expressions, range }
  }

  pub fn get_range(&self) -> Range {
    self.range.clone()
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Identifier {
  pub name: String,
  pub range: Range,
}

impl Identifier {
  pub fn new(name: String, range: Range) -> Self {
    Identifier { name, range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CallExpression {
  pub name: Token,
  pub args: Box<Expression>,
  pub range: Range,
}

impl CallExpression {
  pub fn new(name: Token, args: Box<Expression>, range: Range) -> Self {
    CallExpression { name, args, range }
  }

  pub fn get_range(&self) -> Range {
    let left_range = self.name.range.clone();
    let right_range = self.args.get_range();
    create_middle_range(&left_range, &right_range)
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UnaryExpression {
  pub range: Range,
  pub operator: UnaryOperator,
  pub operand: Box<Expression>,
}

impl UnaryExpression {
  pub fn new(operator: UnaryOperator, operand: Box<Expression>, range: Range) -> Self {
    UnaryExpression { operator, operand, range }
  }

  pub fn get_range(&self) -> Range {
    let left_range = self.operand.get_range();
    let right_range = self.get_operator_range();
    create_middle_range(&left_range, &right_range)
  }

  pub fn get_operator_range(&self) -> Range {
    self.range.clone()
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct FunctionExpression {
  pub arguments: Vec<(Token, Option<Type>)>,
  pub return_type: Option<Type>,
  pub body: Box<Statement>,
  pub range: Range,
}

impl FunctionExpression {
  pub fn new(arguments: Vec<(Token, Option<Type>)>, return_type: Option<Type>, body: Statement, range: Range) -> Self {
    FunctionExpression { arguments, return_type, body: Box::new(body), range }
  }

  pub fn get_range(&self) -> Range {
    let left_range = self.range.clone();
    return left_range;
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BinaryExpression {
  pub operator: BinaryOperator,
  pub left: Box<Expression>,
  pub right: Box<Expression>,
  pub range: Range,
}

impl BinaryExpression {
  pub fn new(operator: BinaryOperator, left: Box<Expression>, right: Box<Expression>, range: Range) -> Self {
    BinaryExpression { operator, left, right, range }
  }
  pub fn get_range(&self) -> Range {
    let left_range = self.left.get_range();
    let right_range = self.right.get_range();
    create_middle_range(&left_range, &right_range)
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum LiteralExpression {
  Number(NumberLiteral),
  String(StringLiteral),
  Boolean(BooleanLiteral),
  Nil,
}

impl LiteralExpression {
  pub fn new_number(value: String, range: Range) -> Self {
    LiteralExpression::Number(NumberLiteral::new(value, range))
  }

  pub fn new_string(value: String, range: Range) -> Self {
    LiteralExpression::String(StringLiteral::new(value, range))
  }

  pub fn new_bool(value: bool, range: Range) -> Self {
    LiteralExpression::Boolean(BooleanLiteral::new(value, range))
  }

  pub fn get_range(&self) -> Range {
    match self {
      LiteralExpression::Number(number) => number.range.clone(),
      LiteralExpression::String(string) => string.range.clone(),
      LiteralExpression::Boolean(boolean) => boolean.range.clone(),
      LiteralExpression::Nil => Range::new(),
    }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct NumberLiteral {
  pub value: String,
  pub range: Range,
}

impl NumberLiteral {
  pub fn new(value: String, range: Range) -> Self {
    NumberLiteral { value, range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct StringLiteral {
  pub value: String,
  pub range: Range,
}

impl StringLiteral {
  pub fn new(value: String, range: Range) -> Self {
    StringLiteral { value, range }
  }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BooleanLiteral {
  pub value: bool,
  pub range: Range,
}

impl BooleanLiteral {
  pub fn new(value: bool, range: Range) -> Self {
    BooleanLiteral { value, range }
  }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum UnaryOperator {
  Negate,
  Not,
  Hash,
}

impl UnaryOperator {
  pub fn to_str(&self) -> &str {
    match self {
      UnaryOperator::Negate => "-",
      UnaryOperator::Not => "not",
      UnaryOperator::Hash => "#",
    }
  }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum BinaryOperator {
  Add,                // +
  Subtract,           // -
  Multiply,           // *
  Divide,             // /
  Modulus,            // %
  And,                // and
  Or,                 // or
  Equal,              // ==
  NotEqual,           // ~=
  LessThan,           // <
  GreaterThan,        // >
  LessThanOrEqual,    // <=
  GreaterThanOrEqual, // >=
  DoubleDot,          // ..
  DoubleSlash,        // //
}

pub struct TypeExpression {
  pub name: Type,
  pub range: Range,
}

impl BinaryOperator {
  pub fn to_str(&self) -> &str {
    match self {
      BinaryOperator::Add => "+",
      BinaryOperator::Subtract => "-",
      BinaryOperator::Multiply => "*",
      BinaryOperator::Divide => "/",
      BinaryOperator::Modulus => "%",
      BinaryOperator::And => "and",
      BinaryOperator::Or => "or",
      BinaryOperator::Equal => "==",
      BinaryOperator::NotEqual => "~=",
      BinaryOperator::LessThan => "<",
      BinaryOperator::GreaterThan => ">",
      BinaryOperator::LessThanOrEqual => "<=",
      BinaryOperator::GreaterThanOrEqual => ">=",
      BinaryOperator::DoubleDot => "..",
      BinaryOperator::DoubleSlash => "//",
    }
  }
}
