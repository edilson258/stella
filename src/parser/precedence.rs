use crate::ast::tokens::TokenKind;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Precedence {
  Assignment,
  Or,
  And,
  Equality,
  Comparison,
  Term,
  Factor,
  Unary,
  Primary,
  Concat,
}

impl Precedence {
  pub fn operators(self) -> &'static [TokenKind] {
    match self {
      Precedence::Or => &[TokenKind::Or],
      Precedence::And => &[TokenKind::And],
      Precedence::Equality => &[TokenKind::Equal, TokenKind::NotEqual],
      Precedence::Comparison => &[TokenKind::Less, TokenKind::LessEqual, TokenKind::Greater, TokenKind::GreaterEqual],
      Precedence::Term => &[TokenKind::Plus, TokenKind::Minus],
      Precedence::Concat => &[TokenKind::DoubleDot],
      Precedence::Factor => &[TokenKind::Star, TokenKind::Slash, TokenKind::Percent],
      _ => &[], // no binary operators for unary and primary precedence :(
    }
  }

  pub fn next(self) -> Precedence {
    match self {
      Precedence::Assignment => Precedence::Or,
      Precedence::Or => Precedence::And,
      Precedence::And => Precedence::Equality,
      Precedence::Equality => Precedence::Comparison,
      Precedence::Comparison => Precedence::Concat,
      Precedence::Concat => Precedence::Term,
      Precedence::Term => Precedence::Factor,
      Precedence::Factor => Precedence::Unary,
      Precedence::Unary => Precedence::Primary,
      Precedence::Primary => Precedence::Primary,
    }
  }
}
