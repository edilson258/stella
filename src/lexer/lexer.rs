use crate::ast::tokens::{Token, TokenKind};
use crate::diagnostics::report::report_and_exit;
use crate::utils::match_number;
use crate::utils::range::Range;

pub struct Lexer<'a> {
  raw: &'a str,
  column: usize,
  line: usize,
  cursor: usize,
  range_start: usize,
  peeked_token: Option<Token>,
  pub file_name: &'a str,
}

impl<'a> Lexer<'a> {
  pub fn new(raw: &'a str, file_name: &'a str) -> Lexer<'a> {
    Lexer { raw, column: 0, range_start: 0, line: 1, cursor: 0, peeked_token: None, file_name }
  }

  pub fn peek_token(&mut self) -> Token {
    if self.peeked_token.is_none() {
      self.peeked_token = Some(self.read_next_token());
    }
    self.peeked_token.clone().unwrap()
  }

  pub fn next_token(&mut self) -> Token {
    if let Some(token) = self.peeked_token.take() {
      return token;
    }
    self.read_next_token()
  }

  fn read_next_token(&mut self) -> Token {
    self.skip_whitespace();
    self.update_current_range();
    if self.is_end() {
      return Token::new(TokenKind::EOF, self.create_range());
    }

    let current_char = self.peek_one();
    match current_char {
      '+' => self.read_check_ahead("+=", TokenKind::Plus, TokenKind::PlusAssign),
      '-' => self.read_comment_or_minus(),
      '*' => self.read_check_ahead("*=", TokenKind::Star, TokenKind::StarAssign),
      '=' => self.read_check_ahead("==", TokenKind::Assign, TokenKind::Equal),
      '~' => self.read_check_ahead("~=", TokenKind::Tilde, TokenKind::NotEqual),
      '<' => self.read_check_ahead("<=", TokenKind::Less, TokenKind::LessEqual),
      '>' => self.read_check_ahead(">=", TokenKind::Greater, TokenKind::GreaterEqual),
      '(' => self.read_simple_token(TokenKind::LeftParen),
      ')' => self.read_simple_token(TokenKind::RightParen),
      '%' => self.read_simple_token(TokenKind::Percent),
      '#' => self.read_simple_token(TokenKind::Hash),
      ',' => self.read_simple_token(TokenKind::Comma),
      ':' => self.read_simple_token(TokenKind::Colon),
      ';' => self.read_simple_token(TokenKind::Semicolon),
      '{' => self.read_simple_token(TokenKind::LeftBrace),
      '}' => self.read_simple_token(TokenKind::RightBrace),
      '[' => self.read_simple_token(TokenKind::LeftBracket),
      ']' => self.read_simple_token(TokenKind::RightBracket),
      '"' => self.read_string(),
      '\'' => self.read_single_quote_string(),
      '/' => self.read_slash(),
      '0'..='9' => self.read_number(),
      'a'..='z' | 'A'..='Z' | '_' => self.read_keyword_or_identifier(),
      '.' => self.read_dot(),
      _ => {
        let mut range = self.create_range();
        let message = format!("Invalid character '{}'", current_char);
        report_and_exit(&message, &mut range, &self.raw, &self.file_name);
      }
    }
  }

  fn read_dot(&mut self) -> Token {
    if self.starts_with("...") {
      self.advance_many(3);
      return Token::new(TokenKind::TripleDot, self.create_range());
    } else {
      self.read_check_ahead("..", TokenKind::Dot, TokenKind::DoubleDot)
    }
  }

  fn read_slash(&mut self) -> Token {
    if self.starts_with("//") {
      self.read_check_ahead("//", TokenKind::DoubleSlash, TokenKind::DoubleSlash)
    } else {
      self.read_check_ahead("/=", TokenKind::Slash, TokenKind::SlashAssign)
    }
  }

  fn read_simple_token(&mut self, kind: TokenKind) -> Token {
    self.advance_one();
    let range = self.create_range();
    Token::new(kind, range)
  }

  fn read_check_ahead(&mut self, expected: &str, single_kind: TokenKind, double_kind: TokenKind) -> Token {
    if self.starts_with(expected) {
      self.advance_many(expected.len());
      let range = self.create_range();
      return Token::new(double_kind, range);
    }
    self.read_simple_token(single_kind)
  }

  fn read_comment_or_minus(&mut self) -> Token {
    if self.starts_with("--") {
      self.read_comment()
    } else {
      self.read_check_ahead("-=", TokenKind::Minus, TokenKind::MinusAssign)
    }
  }

  fn read_comment(&mut self) -> Token {
    if self.starts_with("--[[") {
      self.read_block_comment()
    } else {
      self.read_line_comment()
    }
  }

  fn read_block_comment(&mut self) -> Token {
    self.consume_expect("--[[");
    let text = self.read_until("--]]");
    self.consume_expect_with_custom_error("--]]", "unexpected end of block comment");
    let range = self.create_range();
    Token::new_block_comment(range, text)
  }

  fn read_line_comment(&mut self) -> Token {
    self.consume_expect("--");
    let text = self.read_while(|c| c != '\n');
    let range = self.create_range();
    Token::new_comment(range, text)
  }

  fn read_keyword_or_identifier(&mut self) -> Token {
    let text = self.read_while(|c| c.is_ascii_alphabetic() || c == '_' || c == '$' || c.is_ascii_digit());
    let range = self.create_range();
    Token::new_keyword_or_identifier(range, text)
  }

  fn read_number(&mut self) -> Token {
    let number = self.read_while(match_number);
    let range = self.create_range();
    Token::new_number(range, number)
  }

  fn read_string(&mut self) -> Token {
    self.consume_expect("\"");
    let string = self.read_while(|c| c != '"');
    self.consume_expect_with_custom_error("\"", "unterminated string");
    let range = self.create_range();
    Token::new_string(range, string)
  }

  fn read_single_quote_string(&mut self) -> Token {
    self.consume_expect("'");
    let string = self.read_while(|c| c != '\'');
    self.consume_expect_with_custom_error("'", "unterminated string");
    let range = self.create_range();
    Token::new_string(range, string)
  }

  fn read_while(&mut self, mut test: impl FnMut(char) -> bool) -> String {
    let range_start = self.cursor;
    while !self.is_end() && test(self.peek_one()) {
      self.advance_one();
    }
    self.raw[range_start..self.cursor].to_string()
  }

  fn read_until(&mut self, end: &str) -> String {
    let range_start = self.cursor;
    while !self.is_end() && !self.starts_with(end) {
      let c = self.peek_one();
      if c == '\n' {
        self.advance_new_line();
      } else {
        self.advance_one();
      }
    }
    self.raw[range_start..self.cursor].to_string()
  }

  fn advance_one(&mut self) {
    if let Some(c) = self.raw[self.cursor..].chars().next() {
      self.cursor += c.len_utf8();
      self.column += 1;
      if c == '\n' {
        self.line += 1;
        self.column = 0;
      }
    }
  }

  fn create_range(&mut self) -> Range {
    let start = self.range_start;
    let end = self.cursor;
    self.range_start = self.cursor;
    Range { start, end }
  }

  fn update_current_range(&mut self) {
    self.range_start = self.cursor;
  }

  fn consume_expect(&mut self, text: &str) {
    if self.starts_with(text) {
      self.advance_many(text.len());
    } else {
      let range = self.create_range();
      panic!("expected '{}' at {:?}", text, range);
    }
  }

  fn consume_expect_with_custom_error(&mut self, text: &str, error_message: &str) {
    if self.starts_with(text) {
      self.advance_many(text.len());
    } else {
      let location = self.create_range();
      panic!("{} at {:?}", error_message, location);
    }
  }

  fn is_end(&self) -> bool {
    self.cursor >= self.raw.len()
  }

  fn peek_one(&self) -> char {
    self.raw[self.cursor..].chars().next().unwrap_or('\0')
  }

  fn advance_many(&mut self, count: usize) {
    for _ in 0..count {
      self.advance_one();
    }
  }

  fn starts_with(&self, s: &str) -> bool {
    self.raw[self.cursor..].starts_with(s)
  }

  fn advance_new_line(&mut self) {
    self.line += 1;
    self.column = 0;
    self.cursor += 1;
  }

  fn skip_whitespace(&mut self) {
    while !self.is_end() && self.peek_one().is_whitespace() {
      if self.peek_one() == '\n' {
        self.advance_new_line();
      } else {
        self.advance_one();
      }
    }
  }
}
