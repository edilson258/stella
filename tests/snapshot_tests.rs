#![allow(dead_code, unused_imports)]

use glob::glob;
use insta::assert_ron_snapshot;
use std::ffi::OsString;
use std::fs;
use std::path::Path;
use stella::ast::tokens::{Token, TokenKind};
use stella::lexer::Lexer;
use stella::parser::parser::Parser;

fn read_test_files_with_pattern(pattern: &str) -> Vec<(String, String)> {
  let mut patterns = Vec::new();
  let glob_pattern = glob(pattern).expect("Failed to read glob pattern");
  for entry in glob_pattern {
    let path = entry.expect("Failed to read file");
    let file_name = path.file_name().unwrap().to_string_lossy().to_string();
    let content = fs::read_to_string(path).expect("Failed to read file");
    patterns.push((file_name.to_owned(), content));
  }
  patterns
}

fn create_tokens(source_code: &str) -> Vec<Token> {
  let mut lexer = Lexer::new(source_code.to_string());
  let mut tokens = vec![];
  loop {
    let token = lexer.next_token();
    if token.kind == TokenKind::EOF {
      tokens.push(token);
      break;
    }
    tokens.push(token);
  }
  return tokens;
}

fn create_parser(source_code: &str) -> Parser {
  let parser = Parser::new(source_code);
  return parser;
}
fn format_file_name_with_module(file_name: &str, module: &str) -> String {
  let file_name = format!("{}_{}", module, file_name).replace(".lua", "");
  return file_name;
}
fn setings_snapshot() -> insta::Settings {
  let mut settings = insta::Settings::clone_current();
  settings.set_prepend_module_to_snapshot(false);
  settings.set_omit_expression(true);
  settings
}

#[test]
fn test_lexer_snapshot() {
  let test_files = read_test_files_with_pattern("tests/golden_tests/*.lua");
  let settings = setings_snapshot();
  settings.bind(|| {
    for (file_name, source_code) in test_files.iter() {
      let tokens = create_tokens(source_code);
      let file_name = format_file_name_with_module(file_name, "lexer");
      assert_ron_snapshot!(file_name.clone(), tokens);
    }
  });
}

#[test]
fn test_parser_snapshot() {
  let test_files = read_test_files_with_pattern("tests/golden_tests/*.lua");
  let settings = setings_snapshot();
  settings.bind(|| {
    for (file_name, source_code) in test_files.iter() {
      let mut parser = create_parser(source_code);
      let file_name = format_file_name_with_module(file_name, "parser");
      let program = parser.parse_program();
      assert_ron_snapshot!(file_name.clone(), program);
    }
  });
}

#[test]
fn test_type_checker_snapshot() {
  let test_files = read_test_files_with_pattern("tests/golden_tests/*.lua");
  println!("test_type_checker_snapshot: {:?}", test_files);
  // hei, please :) implement me
}
