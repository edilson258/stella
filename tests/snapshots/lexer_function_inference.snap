---
source: tests/snapshot_tests.rs
---
[
  Token(
    kind: Function,
    range: Range(
      start: 0,
      end: 8,
    ),
  ),
  Token(
    kind: Identifier("add"),
    range: Range(
      start: 9,
      end: 12,
    ),
  ),
  Token(
    kind: LeftParen,
    range: Range(
      start: 12,
      end: 13,
    ),
  ),
  Token(
    kind: Identifier("a"),
    range: Range(
      start: 13,
      end: 14,
    ),
  ),
  Token(
    kind: Comma,
    range: Range(
      start: 14,
      end: 15,
    ),
  ),
  Token(
    kind: Identifier("b"),
    range: Range(
      start: 16,
      end: 17,
    ),
  ),
  Token(
    kind: RightParen,
    range: Range(
      start: 17,
      end: 18,
    ),
  ),
  Token(
    kind: Return,
    range: Range(
      start: 23,
      end: 29,
    ),
  ),
  Token(
    kind: Identifier("a"),
    range: Range(
      start: 30,
      end: 31,
    ),
  ),
  Token(
    kind: Plus,
    range: Range(
      start: 32,
      end: 33,
    ),
  ),
  Token(
    kind: Identifier("b"),
    range: Range(
      start: 34,
      end: 35,
    ),
  ),
  Token(
    kind: End,
    range: Range(
      start: 36,
      end: 39,
    ),
  ),
  Token(
    kind: Local,
    range: Range(
      start: 41,
      end: 46,
    ),
  ),
  Token(
    kind: Identifier("result"),
    range: Range(
      start: 47,
      end: 53,
    ),
  ),
  Token(
    kind: Assign,
    range: Range(
      start: 54,
      end: 55,
    ),
  ),
  Token(
    kind: Identifier("add"),
    range: Range(
      start: 56,
      end: 59,
    ),
  ),
  Token(
    kind: LeftParen,
    range: Range(
      start: 59,
      end: 60,
    ),
  ),
  Token(
    kind: Number("5"),
    range: Range(
      start: 60,
      end: 61,
    ),
  ),
  Token(
    kind: Comma,
    range: Range(
      start: 61,
      end: 62,
    ),
  ),
  Token(
    kind: Number("10"),
    range: Range(
      start: 63,
      end: 65,
    ),
  ),
  Token(
    kind: RightParen,
    range: Range(
      start: 65,
      end: 66,
    ),
  ),
  Token(
    kind: EOF,
    range: Range(
      start: 67,
      end: 67,
    ),
  ),
]
